src/url.rs

```diff
@@ -48,48 +48,38 @@ pub struct HttpUrl {
 }
 
 impl HttpUrl {
-    pub fn port(&self) -> Result<u16> {
-        match self.url.port_or_known_default() {
-            Some(v) => Ok(v),
-            None => Err(Error::UrlError(format!(
-                "unsupported URL scheme {}: no port and no known default",
-                self.url.scheme()
-            ))),
-        }
+    pub fn port(&self) -> u16 {
+        // this should always go to the happy path as we only have scheme of http and https
+        self.url.port_or_known_default().unwrap()
     }
     pub fn scheme(&self) -> Scheme {
         self.scheme.clone()
     }
     pub fn host(&self) -> &str {
         &self.host
     }
-    // give use an option to use the origin url struct and save us from porting too many codes
     pub fn url(&self) -> &Url {
         &self.url
     }
 }
 
+#[inline]
+fn error_unsupported_url_scheme(scheme: &str) -> Error {
+    Error::UrlError(format!("unsupported URL scheme {}", scheme))
+}
+
 impl TryFrom<Url> for HttpUrl {
     type Error = Error;
 
     fn try_from(url: Url) -> Result<Self> {
         use core::str::FromStr;
 
         let scheme = Scheme::from_str(url.scheme())?;
-        if scheme.ne(&Scheme::Http) && scheme.ne(&Scheme::Https) {
-            return Err(Error::UrlError(format!(
-                "unsupported URL scheme {}",
-                url.scheme()
-            )));
-        };
-        let Some(host) = url.host_str() else {
-            // people can use url.set_host(Option::None) to make host invalid, and currently we
-            // don't support that
-            return Err(Error::UrlError(format!(
-                "unsupported URL scheme {}: no host",
-                url.scheme()
-            )));
+        if scheme != Scheme::Http && scheme != Scheme::Https {
+            return Err(error_unsupported_url_scheme(url.scheme()));
         };
+        // host must exist in http and https, and url crate ensure about it, see test check_url_must_have_host
```

Comment:
I would phrase this as "HTTP and HTTPS URLs must always have a host, see the check_url_must_have_host test"